smithy-model/src/main/java/software/amazon/smithy/model/validation/ValidationEventDecoratorFactory.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.validation;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * Creates {@link ValidationEventDecorator} instances.
+ */
+public interface ValidationEventDecoratorFactory {
```

Comment:
That was my first take on this (e.g., the first [commit](https://github.com/awslabs/smithy/pull/1774/commits/110b1010d3826616fe5359e28b51d3131c7284f6#diff-fa50f59b51956a1b4f6c0286be3eab280dd0eae39c95ed753ac0c3437c07e466R42-R50) of this PR), however some of the validation events that we're interested in (e.g., undefined trait), are returned earlier [here](https://github.com/awslabs/smithy/blob/7342b2c5168dc868c49241e22632a18ca63438bb/smithy-model/src/main/java/software/amazon/smithy/model/loader/ModelAssembler.java#L574-L576) before the validation phase, thus I decoupled the decoration from the validation phase.

Reply:
I'm not sure that I follow how that impacts what creates the decorators.

Reply:
It does not, we could use the `ValidationFactory`, it's just that we might need to pull up its default value into the assembler instead of having it inside the `ModelValidator` class [here](https://github.com/awslabs/smithy/blob/7342b2c5168dc868c49241e22632a18ca63438bb/smithy-model/src/main/java/software/amazon/smithy/model/loader/ModelValidator.java#L167-L169), and then pass it down. I can make that change.